### PR TITLE
[Security] Improved error message when no supported user provider is found

### LINKED ISSUE #35435: [security] There is no user provider for user
**Symfony version(s) affected**: 4.4.3

**Description**  
After update to 4.4.3 i have a RuntimeException "There is no user provider for user"

**How to reproduce**  
if you have custom provider update to 4.4.3

Sample
```yml
security:
    providers:
        custom-security:
            id: custom.user_provider
    firewalls:
        dev:
            pattern: ^/(_(profiler|wdt)|css|images|js)/
            security: false
        login:
            pattern: ^/login
            security: false
        secured_area:
            pattern: ^/
            guard:
                authenticators:
                    - custom.authentificator
                entry_point: custom.authentificator
            logout:
                path: /logout
```

**Possible Solution**  
Problem is on the file  Symfony/Component/Security/Http/Firewall/ContextListener.php method refreshUser

### PR DESCRIPTION
**Pull request by wouterj**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #35435
| License       | MIT
| Doc PR        | -

### CODE REVIEW

File: src/Symfony/Component/Security/Core/User/ChainUserProvider.php [hunk 1]
```diff
@@ -91,7 +91,7 @@ public function refreshUser(UserInterface $user)
             $e->setUsername($user->getUsername());
             throw $e;
         } else {
-            throw new UnsupportedUserException(sprintf('The account "%s" is not supported.', \get_class($user)));
+            throw new UnsupportedUserException(sprintf('There is no user provider for user "%s.%s', \get_class($user), count($this->providers) > 0 ? ' Are you sure supportsClass() of your user provider returns true for this classname?' : ''));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/User/ChainUserProvider.php]: missing double quote after `%s`
+
`Shouldn't the "supportsClass()" method of your user provider returns true for this classname?`
same below

Comment (chalasr) [reply src/Symfony/Component/Security/Core/User/ChainUserProvider.php]: 👍 at this stage we can safely assume that a broken  `supportsClass` implementation is the cause

Comment (chalasr) [reply src/Symfony/Component/Security/Core/User/ChainUserProvider.php]: 👍 at this stage we can safely assume that a broken supportsClass implementation is the cause

File: src/Symfony/Component/Security/Http/Firewall/ContextListener.php [hunk 2]
```diff
@@ -292,7 +292,7 @@ protected function refreshUser(TokenInterface $token)
             return null;
         }
 
-        throw new \RuntimeException(sprintf('There is no user provider for user "%s".', $userClass));
+        throw new UnsupportedUserException(sprintf('There is no user provider for user "%s.%s', $userClass, count($this->userProviders) > 0 ? ' Are you sure supportsClass() of your user provider returns true for this classname?' : ''));
```
Comment (chalasr) [inline src/Symfony/Component/Security/Http/Firewall/ContextListener.php]: Reaching this exception means you have to fix your code, a `\LogicException` would be a better a fit. 
Also we should double check that this change does not impact the current error handling given it's an `AuthenticationServiceException` (is there anything likely to catch it and perform some special treatment regarding e.g. logging or response handling?)

Comment (wouterj) [reply src/Symfony/Component/Security/Http/Firewall/ContextListener.php]: Oh, I just was a lazy copy paster and didn't even notice this used a different class. I've reverted it back to `\RuntimeException`, it wasn't my intention to change it.

File: src/Symfony/Component/Security/Http/Firewall/ContextListener.php [hunk 3]
```diff
@@ -292,7 +292,7 @@ protected function refreshUser(TokenInterface $token)
             return null;
         }
 
-        throw new \RuntimeException(sprintf('There is no user provider for user "%s".', $userClass));
+        throw new \RuntimeException(sprintf('There is no user provider for user "%s".%s', \get_class($user), count($this->providers) > 0 ? ' Shouldn\'t the "supportsClass()" method of your user provider returns true for this classname?' : ''));
```
Comment (chalasr) [inline src/Symfony/Component/Security/Http/Firewall/ContextListener.php]: `returns` => `return`?

**Comment by wouterj** [comment]
Thanks for the quick review (and much nicer wording!), should be ready now.

File: src/Symfony/Component/Security/Http/Firewall/ContextListener.php (hunk 3 above)
Comment (wouterj) [reply src/Symfony/Component/Security/Http/Firewall/ContextListener.php]: Isn't "the .. method of your user provider" third person? I agree that "returns" sounds a bit weird, but I feel like gramatically it's more correct. (but I'm not close to native english)

Comment (chalasr) [reply src/Symfony/Component/Security/Http/Firewall/ContextListener.php]: I think the verb-subject agreement applies to `should` here (can be replaced by `does it return?`.)
More importantly: shouldn't `$this->providers` be `$this->userProviders` here? 
Also `iterable` is not always count-able. Let's skip the check, you won't reach this statement if `$this->providers <= 0` anyway, same for other patched classes.
>

Comment (wouterj) [reply src/Symfony/Component/Security/Http/Firewall/ContextListener.php]: Oh, alright. Updated all 3 cases

**Review by chalasr — APPROVED** [review]

**Comment by chalasr** [comment]
This applies to 3.4 isn't it?

**Comment by wouterj** [comment]
Oh sorry, yes should be 3.4

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @wouterj.

**Pull request merged**